### DISCUSSION TITLE
Is inner axon space mask necessary?

### DISCUSSION THREAD

**Discussion by akraskovSMDB** [ivadomed/ivadomed#1293] [General]
Hi, 

I am trying to understand a little bit more about the ivadomed. I use it in combination with axondeepseg for the morphological analysis of axons. 

A training flow assumes two manually segmented masks: myelin and axonal inner space.  But these two masks are not independent, in the sense that the axonal inner space mask can be reconstructed from myelin (almost always except for the cases with many (>2) particularly touching axons).

Does ivadomed use information about two masks jointly or independently?
I understand that both masks are used simultaneously during the model training, my question is more, does ivadomed take into account that the axonal mask for each axon is within myelin mask and that they are touching?
Does the axonal mask actually add anything to the performance of the segmentation, or myelin mask would be enough?
In terms of contrast differences axonal mask is more similar to the outside space, does this cause any problems for model training?

A second question (it is related but happy to move to a different topic).
Assuming that axonal mask is not essential, and only myelin mask is important, the segmentation problem is reduced to two classes (myelin and everything else) and it can be tackled using standard threshold-based methods (e.g. matlab Image Segmenter App a recent toy to play with without a need to programme anything). Does ivadomed still provide better segmentation results? Did anyone recently try to do a comprehensive comparison between standard and ivadomed?

I do appreciate the universality of ivadomed with respect to the image contrast. On another side, one can probably build in some image pre-processing tools (has been raised before see https://github.com/ivadomed/ivadomed/issues/26). 

Thank you

**Comment by hermancollin** [comment]
Hi @akraskovSMDB.
> Does ivadomed use information about two masks jointly or independently?

The ivadomed model has 2 output channels, one for the myelin and one for the axon. During the training, the loss is computed on both masks. You could train a model to predict only the myelin if you want.

>  the segmentation problem is reduced to two classes (myelin and everything else) and it can be tackled using standard threshold-based methods (e.g. matlab Image Segmenter App a recent toy to play with without a need to programme anything). Does ivadomed still provide better segmentation results? Did anyone recently try to do a comprehensive comparison between standard and ivadomed?

Even the original problem with both axon and myelin masks could be tackled using traditional image processing. However, there's extensive litterature on the image segmentation task and deep learning based methods outperform traditional machine learning on the harder problems. At the end of the day, you can use any tool you want to obtain your segmentations. Deep Learning offers an efficient solution for robust automated segmentation but other methods exists, although they are not as popular nowadays. With that being said, I know some people still use traditional ML algorithms for segmentation.

If you are completely new to deep learning, I would advise against trying to train your own models with ivadomed. It is no longer active - on the `axondeepseg` side we are exploring alternatives. If you have a small amount of images, it might be easier with other tools (like this matlab module you are referring to).

May I ask why you are training models for axon/myelin segmentation? Have you tried some of the pre-trained axondeepseg models?

**Reply by akraskovSMDB** [reply]
Hi @hermancollin , 

> May I ask why you are training models for axon/myelin segmentation? Have you tried some of the pre-trained axondeepseg models?

Standard models neither SEM nor TEM give me any reasonable results. See the images at the end. Do you have other standard models which I can try on the image shown below.
How many images (or better axons) went into the standard TEM model?

I am not completely new to deep learning, it is just when I use a more sophisticated method, especially a block box one (like CNN) I'd like to be sure that a more simple and intuitive one (like simple thresholding-based methods) does not give the same results. 

> However, there's extensive litterature on the image segmentation task and deep learning based methods outperform traditional machine learning on the harder problems.

Could you please give me a couple of references in relation to the segmentation problem (in addition to Zaimi et al)

> You could train a model to predict only the myelin if you want

How do I do this exactly? Do I just supply myelin masks within BIDS structure and remove axon mask? 

Does ivadomed take into account that the axonal mask for each axon is within myelin mask and that they are touching?

How close you are in your search for alternatives for ivadomed?

My sample EM image (resolution 0.026um/pix) and results of segmentation by standard SEM and TEM models  

![hEM20229919-bA-S1-32](https://github.com/ivadomed/ivadomed/assets/25688017/be65ae20-7de5-43a1-9966-bc36e892101d)
![hEM20229919-bA-S1-32_seg-axonmyelin-SEM](https://github.com/ivadomed/ivadomed/assets/25688017/30915b7d-d6c6-43bd-b8d8-788dfc46eec6)
![hEM20229919-bA-S1-32_seg-axonmyelin-TEM](https://github.com/ivadomed/ivadomed/assets/25688017/27a2b24d-8f2e-4cb0-9a41-5a635daad295)

**Reply by hermancollin** [reply]
> How many images (or better axons) went into the standard TEM model?

158 annotated images went into the standard TEM model training. In terms of axons, it's very hard to tell but I would say between 5k and 10k.
 
> > However, there's extensive litterature on the image segmentation task and deep learning based methods outperform traditional machine learning on the harder problems.
> 
> Could you please give me a couple of references in relation to the segmentation problem (in addition to Zaimi et al)

Better yet you can take a look at the benchmarks here: https://paperswithcode.com/task/medical-image-segmentation. These benchmarks are specific to medical image segmentation. To my knowledge, all the best models listed are deep learning architectures. (papers are listed)
 
> > You could train a model to predict only the myelin if you want
> 
> How do I do this exactly? Do I just supply myelin masks within BIDS structure and remove axon mask?

Yes exactly.

> Does ivadomed take into account that the axonal mask for each axon is within myelin mask and that they are touching?

No but this should be the case by default if the annotations are as expected.

> How close you are in your search for alternatives for ivadomed?

Well in fact I am already testing an alternative for the backbone of ADS. The reason why the ivadomed software is no longer active is that it was costly to maintain and alternatives like `monai` are very popular in the community (with big contributors such as NVIDIA).

> Standard models neither SEM nor TEM give me any reasonable results. See the images at the end. Do you have other standard models which I can try on the image shown below.

Hmm that's weird. Did you try the ADS zoom factor feature? I quickly ran your image in ADS with `zoom_factor=0.4` and I got this pretty clean result for a first try. Results could be even more improved by fine-tuning the zoom factor parameter further.
![242370736-be65ae20-7de5-43a1-9966-bc36e892101d_seg-axonmyelin](https://github.com/ivadomed/ivadomed/assets/83031821/ddb4ab9c-024b-4100-88d0-f3abf293ee2e)
Your image resolution of 0.026 and the model was trained on 0.01 so some zoom factor adjustment is important to optimize the results.

**Reply by akraskovSMDB** [reply]
sorry, @hermancollin, I am confused about -s (pixel size) parameter and zoom factor, and they interact, I thought that pixel size parameter is used to adjust the image to the model. Are you saying that both parameters matter and I have to use choose -s and -z in such a way that s*z~0.01? Will the result be the same if I choose -s 0.26 and zoom 0.04?
Was your segmentation done with --no-patch option?

>>>You could train a model to predict only the myelin if you want

>>How do I do this exactly? Do I just supply myelin masks within BIDS structure and remove axon mask?

>Yes exactly.

But in this case, I will still need somehow estimate the axonal inner space mask since axondeepseg_morphometrics will not work without it, correct? Or do you have tools which can create axon inner space mask given myelin mask?

**Reply by hermancollin** [reply]
> sorry, @hermancollin, I am confused about -s (pixel size) parameter and zoom factor, and they interact, I thought that pixel size parameter is used to adjust the image to the model. Are you saying that both parameters matter and I have to use choose -s and -z in such a way that s*z~0.01? Will the result be the same if I choose -s 0.26 and zoom 0.04? Was your segmentation done with --no-patch option?

Yes, apologies, maybe I wasn't clear enough. Our documentation might need to be a little more explicit with the zoom factor feature. I used `-s 0.026` to specify the actual pixel size of the image and `-z 0.4` to specify the adjustment. This gives an "effective" resolution of **0.026 * 0.4 = 0.0104 um/px**, which is close to 0.01 the standard resolution for our TEM model. In your case, `-s 0.26 -z 0.04` will give the same "effective" resolution but be careful because this is not the actual pixel size of the image , which is important when computing morphometrics. 

> > > > You could train a model to predict only the myelin if you want
> 
> > > How do I do this exactly? Do I just supply myelin masks within BIDS structure and remove axon mask?
> 
> > Yes exactly.
> 
> But in this case, I will still need somehow estimate the axonal inner space mask since axondeepseg_morphometrics will not work without it, correct? Or do you have tools which can create axon inner space mask given myelin mask?

Yes it would require minor post-processing. You could do it automatically using basic morphological operations or manually "fill" the axon class in an image editing software. We do have a tool to fill axons quickly in our graphical interface.
